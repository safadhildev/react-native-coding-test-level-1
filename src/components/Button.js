import { TouchableOpacity, StyleSheet } from 'react-native';
import _ from 'lodash';
import Text from './Text';

import styles from './styles';

const buttonStyle = (variant) =>
  _.isEqual(variant, 'outlined') ? styles.outlinedButton : styles.button;
const buttonTextStyle = (variant) =>
  _.isEqual(variant, 'outlined') ? styles.outlinedButtonText : styles.buttonText;

const Button = ({ variant = 'outlined', text = '', disabled, onPress = () => {}, ...props }) => {
  return (
    <TouchableOpacity
      {...props}
      disabled={disabled}
      onPress={onPress}
      style={[buttonStyle(variant), disabled && styles.buttonDisabled]}
      activeOpacity={0.8}
    >
      <Text style={buttonTextStyle(variant)}>{text}</Text>
    </TouchableOpacity>
  );
};

export default Button;
