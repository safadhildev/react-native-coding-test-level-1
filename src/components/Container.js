import { SafeAreaView, View } from 'react-native';
import styles from './styles';

const Container = ({ children }) => {
  return <SafeAreaView style={styles.container}>{children}</SafeAreaView>;
};

export default Container;
