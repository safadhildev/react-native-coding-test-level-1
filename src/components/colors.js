export default {
  black: '#222225',
  green: '#50C878',
  grey: '#757575',
  red: '#D32F2F',
  white: '#F0F0F0',
  teal: '#0097A7',
};
