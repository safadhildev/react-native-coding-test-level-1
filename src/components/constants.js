import routes from '../routes/routes';

const constants = {
  routes,
};

export default constants;
