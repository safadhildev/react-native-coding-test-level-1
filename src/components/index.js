export { default as Button } from './Button';
export { default as Container } from './Container';
export { default as Text } from './Text';
export { default as Input } from './Input';
export { default as DateInput } from './DateInput';
export { default as Header } from './Header';

export { default as commonStyles } from './styles';
