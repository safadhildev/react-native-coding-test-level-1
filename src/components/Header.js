import { TouchableOpacity, View } from 'react-native';
import Text from './Text';
import styles from './styles';
import { useNavigation } from '@react-navigation/native';
import { ArrowLeft } from 'iconoir-react-native';
import colors from './colors';

const Header = ({ title, disableBackAction = false }) => {
  const navigation = useNavigation();

  const onBack = () => {
    navigation.goBack();
  };

  return (
    <View style={styles.headerContainer}>
      {!disableBackAction && (
        <View style={{ marginRight: 10 }}>
          <TouchableOpacity activeOpacity={0.8} onPress={onBack}>
            <ArrowLeft color={colors.black} height={28} width={28} strokeWidth={3} />
          </TouchableOpacity>
        </View>
      )}
      <Text style={styles.title}>{title}</Text>
    </View>
  );
};

export default Header;
