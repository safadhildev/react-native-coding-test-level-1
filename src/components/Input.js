import { useEffect, useRef, useState } from 'react';
import { View, TextInput } from 'react-native';
import Text from './Text';

import styles from './styles';
import _ from 'lodash';
import colors from './colors';

const Input = ({
  placeholder = '',
  value = '',
  editable = true,
  onChangeText = () => {},
  onBlur = () => {},
  error = false,
  errorText = '',
  ...props
}) => {
  const [inputFocused, setInputFocused] = useState(false);

  return (
    <View style={{ marginBottom: 5 }}>
      <View style={styles.inputContainer}>
        <View
          pointerEvents="none"
          style={
            inputFocused || !_.isEmpty(value) ? styles.labelWrapper : styles.placeholderWrapper
          }
        >
          <Text
            style={[
              inputFocused || !_.isEmpty(value) ? styles.labelText : styles.placeholderText,
              { color: inputFocused ? colors.teal : colors.grey },
            ]}
          >
            {placeholder}
          </Text>
        </View>
        <TextInput
          editable={editable}
          allowFontScaling={false}
          onFocus={() => {
            setInputFocused(true);
          }}
          onBlur={() => {
            setInputFocused(false);
            onBlur();
          }}
          value={value}
          onChangeText={onChangeText}
          style={[styles.input, inputFocused && { borderColor: colors.teal }]}
          {...props}
        />
      </View>
      <Text style={[styles.errorText, { position: 'absolute', bottom: 0 }]}>{errorText}</Text>
    </View>
  );
};

export default Input;
