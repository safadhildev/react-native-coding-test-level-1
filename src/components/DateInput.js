import { TouchableOpacity } from 'react-native';
import DateTimePicker from '@react-native-community/datetimepicker';
import _ from 'lodash';

import Input from './Input';
import { useState } from 'react';
import moment from 'moment';

const DateInput = ({ value = new Date(), placeholder = '', setValue = () => {} }) => {
  const [showDatePicker, setShowDatePicker] = useState(false);

  const onChange = (event, selectedDate) => {
    if (_.isEqual(event.type, 'set')) {
      setValue(selectedDate);
    }
    setShowDatePicker(false);
  };

  const onShowDatePicker = () => {
    setShowDatePicker(true);
  };

  return (
    <>
      <TouchableOpacity onPress={onShowDatePicker} activeOpacity={0.8}>
        <Input
          value={_.isNull(value) ? 'DD/MM/YYYY' : moment(value).format('DD/MM/YYYY')}
          placeholder={placeholder}
          editable={false}
        />
      </TouchableOpacity>
      {showDatePicker && (
        <DateTimePicker
          testID="dateTimePicker"
          value={_.isNull(value) ? new Date() : value}
          mode="date"
          is24Hour={true}
          onChange={onChange}
          maximumDate={new Date()}
        />
      )}
    </>
  );
};

export default DateInput;
