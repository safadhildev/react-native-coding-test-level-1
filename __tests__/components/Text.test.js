import React from 'react';
import renderer from 'react-test-renderer';
import { Text } from '../../src/components';

describe('render test', () => {
  it('render component (Text) correctly', () => {
    const tree = renderer.create(<Text />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
