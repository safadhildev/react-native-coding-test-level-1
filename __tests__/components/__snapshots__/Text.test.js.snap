// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render test render component (Text) correctly 1`] = `
<Text
  allowFontScaling={false}
  style={
    [
      {
        "fontSize": 16,
        "lineHeight": 24,
      },
      undefined,
    ]
  }
/>
`;
